src/Execution/lbsCriterionBase.py

```diff
@@ -46,78 +46,50 @@
 import sys
 import bcolors
 
-from src.Model.lbsRank import Rank
-
+from src.Model.lbsWorkModelBase import WorkModelBase
 
 class CriterionBase:
     __metaclass__ = abc.ABCMeta
     """An abstract base class of optimization criteria for LBAF execution
     """
 
-    def __init__(self, ranks, edges, parameters=None):
+    def __init__(self, work_model, parameters={}):
```

Comment:
danger, could behave not as intended
https://leimao.github.io/blog/Python-Default-Argument-Mutable-Object/

Reply:
Interesting. I did not know this _caveat_. Is there a PEP8 explicitly forbidding it? Otherwise I would leave it as it is, because it expresses the intent of what `parameters` should be, and forces subclass developers to follow that model.

Reply:
PEP 8 is a style guide for Python Code
https://www.python.org/dev/peps/pep-0008/

Default parameter value is something different:
https://web.archive.org/web/20200221224620/http://effbot.org/zone/default-values.htm

You could use annotations in Python, they help to quickly read intentions as well works fine with IDE:
https://realpython.com/lessons/annotations/

```Python
def __init__(self, ranks, edges, parameters: dict = None):
```

Reply:
Okay. I read the article you sent above, but I still do not see clearly **what** problem it could cause exactly in our case. I would really like to understand what could go wrong here. Do you have an example in mind? Really curious here.

Reply:
In this case maybe when called second time and more it could keep the `old` parameters and add something there instead initiate a new emty dictionary when called.

Reply:
For instance, would it be a good thing to do it in the superclass (the abstract one)?

Like, have `None` passed to superclass but initialize it to `{}` there if it's `None`?

Reply:
No no, please take a look how to deal with that:
```Python
def myfunc(value=None):
    if value is None:
        value = []
```

Reply:
Yes I understand that but I find this ugly.
This is really a flaw of the language that it allows you to define a default value for a function parameter, and yet lets you alter that default inside the function.

Reply:
Philippe it is what it is :), can't change that. Just trying to follow, to get what I want and avoid odd behavior.

Reply:
@marcinwrobel1986 I found a workaround that does not result in having to assign `{}` in case parameters are `None`:
```
        self.dst_work = get_dst_work_know_by_src if not (
            parameters and parameters.get(
                "actual_destination_work")) else get_actual_dst_work
```
and I have combined this with the annotations you suggested. I think the code looks alright like that. Please check it out.

Reply:
@marcinwrobel1986 did you see that change?